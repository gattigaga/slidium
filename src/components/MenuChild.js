import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Icon from "components/Icon";

const Container = styled.div`
  width: 100%;
  padding: 8px 16px;
  box-sizing: border-box;
  cursor: ${({ isClickable }) => (isClickable ? "pointer" : "default")};
  position: relative;
  display: flex;
  align-items: center;
  user-select: none;

  &:hover {
    ${({ isClickable }) => isClickable && "background: #eee;"};
  }
`;

const IconContainer = styled.div`
  width: 32px;
  display: flex;
  position: relative;
`;

const StyledIcon = styled(Icon)`
  font-size: 20px;
  color: ${({ isClickable }) => (isClickable ? "#888" : "#bbb")};
  position: absolute;
  top: 50%;
  left: 0%;
  transform: translateY(-50%);
`;

const Label = styled.span`
  font-size: 12px;
  font-family: Roboto;
  color: ${({ isClickable }) => (isClickable ? "black" : "#bbb")};
`;

const MenuChild = ({ label, icon, isClickable, onClick }) => (
  <Container isClickable={isClickable} onClick={() => isClickable && onClick()}>
    <IconContainer>
      {icon && <StyledIcon name={icon} isClickable={isClickable} />}
    </IconContainer>
    <Label isClickable={isClickable}>{label}</Label>
  </Container>
);

MenuChild.propTypes = {
  icon: PropTypes.string,
  label: PropTypes.string,
  isClickable: PropTypes.bool,
  onClick: PropTypes.func
};

MenuChild.defaultProps = {
  label: "Menu",
  isClickable: true,
  onClick: () => {}
};

export default MenuChild;
