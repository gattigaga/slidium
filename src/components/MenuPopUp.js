import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Container = styled.div`
  position: absolute;
  min-width: 64px;
  top: ${({ position }) => position.y};
  left: ${({ position }) => position.x};
  background: white;
  border: 1px solid #ccc;
  padding: 8px 0px;
  box-shadow: 1px 1px 8px rgba(0, 0, 0, 0.2);
  z-index: 2;
`;

const MenuPopUp = ({ children, position, onClick }) => (
  <Container position={position} onClick={onClick}>
    {children}
  </Container>
);

MenuPopUp.propTypes = {
  children: PropTypes.node.isRequired,
  position: PropTypes.shape({
    x: PropTypes.string,
    y: PropTypes.string
  }),
  onClick: PropTypes.func
};

MenuPopUp.defaultProps = {
  position: {
    x: "0px",
    y: "0px"
  }
};

export default MenuPopUp;
