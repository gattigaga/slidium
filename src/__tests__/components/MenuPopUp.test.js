import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import MenuPopUp from "components/MenuPopUp";
import MenuChild from "components/MenuChild";

let wrapper, onClick;

beforeEach(() => {
  onClick = jest.fn();
  wrapper = shallow(
    <MenuPopUp onClick={onClick}>
      <MenuChild label="New" />
      <MenuChild label="Open" isClickable={false} />
      <MenuChild label="Print" icon="print" />
    </MenuPopUp>
  );
});

describe("MenuPopUp", () => {
  it("renders correctly", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders in other position", () => {
    wrapper = shallow(
      <MenuPopUp>
        <MenuChild label="New" />
        <MenuChild label="Open" isClickable={false} />
        <MenuChild label="Print" icon="print" />
      </MenuPopUp>
    );

    wrapper.setProps({ position: { x: "320px", y: "240px" } });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("calls 'onClick'", () => {
    expect(onClick.mock.calls.length).toEqual(0);
    wrapper.simulate("click");
    expect(onClick.mock.calls.length).toEqual(1);
  });
});
