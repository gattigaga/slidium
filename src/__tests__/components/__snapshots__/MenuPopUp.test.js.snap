// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuPopUp renders correctly 1`] = `
<styled.div
  onClick={[Function]}
  position={
    Object {
      "x": "0px",
      "y": "0px",
    }
  }
>
  <MenuChild
    isClickable={true}
    label="New"
    onClick={[Function]}
  />
  <MenuChild
    isClickable={false}
    label="Open"
    onClick={[Function]}
  />
  <MenuChild
    icon="print"
    isClickable={true}
    label="Print"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`MenuPopUp renders in other position 1`] = `
<styled.div
  position={
    Object {
      "x": "320px",
      "y": "240px",
    }
  }
>
  <MenuChild
    isClickable={true}
    label="New"
    onClick={[Function]}
  />
  <MenuChild
    isClickable={false}
    label="Open"
    onClick={[Function]}
  />
  <MenuChild
    icon="print"
    isClickable={true}
    label="Print"
    onClick={[Function]}
  />
</styled.div>
`;
